### PR TITLE
refactor: avoid wrapping components with injectIntl

### LINKED ISSUE #1412: [regression >3.1.1] Type generics annotation do not get forwarded to FormattedMessage
This commit https://github.com/formatjs/react-intl/commit/660a546a87af198d84d119499c476857e798cac7 changed the exported component `message` component to be wrapped with `injectIntl`.

However, I believe that this change broke the TS generics, as the type of the values field does not get forwarded anymore.

Before I could do:

```tsx
<FormattedMessage<
  Record<string, (msg: React.ReactNodeArray) => JSX.Element>
>
  {...messages.paragraph1}
  values={{
    a: msg => (
      <a
        href={SUPPORT_PORTAL_URL}
        target="_blank"
        rel="noopener noreferrer"
      >
        {msg}
      </a>
    ),
  }}
/>
```

But now it breaks because the generic type that I pass gets ignored

image

Can the component be rewritten using react hooks?

### PR DESCRIPTION
**Pull request by emmenko**
Fixes #1412

TL;DR: I'm not aware of a way to "forward" the types to the wrapped components when using a HOC. To keep things simple, and avoid using the HOC in the first place, we can simply use the context consumer directly.

I also refactored a bit the tests to use `console.error = jest.fn()`, which is the easiest way to mock the log.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -34,15 +34,15 @@
   "module": "./lib/index.js",
   "types": "./dist/index.d.ts",
   "dependencies": {
-    "@formatjs/intl-relativetimeformat": "^2.6.2",
+    "@formatjs/intl-relativetimeformat": "^2.6.3",
```
Comment (emmenko) [inline package.json]: I included those deps updates as I saw they were freshly published. Let me know if it's ok, otherwise I can revert.

File: src/components/html-message.tsx [hunk 2]
```diff
@@ -5,49 +5,59 @@
...
-    // message. It is assumed that the message is not UGC, and came from the
-    // developer making it more like a template.
-    //
-    // Note: There's a perf impact of using this component since there's no
-    // way for React to do its virtual DOM diffing.
-    const html = {__html: formattedHTMLMessage};
-    return <Component dangerouslySetInnerHTML={html} />;
+    return (
+      <Context.Consumer>
+        {intl => {
```
Comment (emmenko) [inline src/components/html-message.tsx]: This is basically everything that we need from the `injectIntl`

File: src/components/message.tsx [hunk 3]
```diff
@@ -5,17 +5,16 @@
...
-const shallowEquals = require('shallow-equal/objects');
-
 import {formatMessage as baseFormatMessage} from '../format';
 import {
   invariantIntlContext,
   DEFAULT_INTL_CONFIG,
   createFormatters,
 } from '../utils';
-import {PrimitiveType, FormatXMLElementFn} from 'intl-messageformat';
+const shallowEquals = require('shallow-equal/objects');
```
Comment (emmenko) [inline src/components/message.tsx]: Just reordering a bit the imports

File: src/components/message.tsx [hunk 4]
```diff
@@ -77,40 +69,50 @@ export class BaseFormattedMessage<
...
-      id,
-      description,
-      defaultMessage,
-      values,
-      children,
-      tagName: Component = Text,
-    } = this.props;
+    return (
+      <Context.Consumer>
+        {intl => {
```
Comment (emmenko) [inline src/components/message.tsx]: Same think here

File: src/components/relative.tsx [hunk 5]
```diff
@@ -177,38 +173,46 @@ export class BaseFormattedRelativeTime extends React.PureComponent<
...
-
-    const formattedRelativeTime = formatRelativeTime(
-      currentValue,
-      currentUnit,
-      {
-        ...this.props,
-      }
+    return (
+      <Context.Consumer>
+        {intl => {
```
Comment (emmenko) [inline src/components/relative.tsx]: Also for this one, no need to have a HOC here

File: test/unit/components/date.tsx [hunk 6]
```diff
@@ -73,12 +68,17 @@ describe('<FormattedDate>', () => {
...
+  it('falls back and warns on invalid Intl.DateTimeFormat options', () => {
     const date = new Date();
     const rendered = mountWithProvider({value: date, year: 'invalid'}, intl);
 
     expect(rendered.text()).toBe(String(date));
-    expect(consoleError.mock.calls.length).toBeGreaterThan(0);
+    expect(console.error).toHaveBeenCalledTimes(1);
+    expect(console.error).toHaveBeenCalledWith(
+      expect.stringContaining(
+        'Error formatting date.\nRangeError: Value invalid out of range for dateformat options property year'
```
Comment (emmenko) [inline test/unit/components/date.tsx]: I added more expectations to check for the error message, as it gives a better understanding of the expected behavior.

File: test/unit/components/provider.tsx [hunk 7]
```diff
@@ -1,9 +1,12 @@
 import * as React from 'react';
 import {mount} from 'enzyme';
+import areIntlLocalesSupported from 'intl-locales-supported';
 import IntlProvider, {Props} from '../../../src/components/provider';
 import {WithIntlProps} from '../../../src/components/injectIntl';
 import {IntlShape} from '../../../src/types';
-import withIntl, {Provider} from '../../../src/components/injectIntl';
+import withIntl from '../../../src/components/injectIntl';
+
+jest.mock('intl-locales-supported');
```
Comment (emmenko) [inline test/unit/components/provider.tsx]: This is necessary otherwise the test fails because the intl polyfill (for the tests) returns `false` for some of the intl contructors as they don't support languages other than `en`.

File: test/unit/components/provider.tsx [hunk 8]
```diff
@@ -93,6 +94,8 @@ describe('<IntlProvider>', () => {
   });
 
   it('provides `context.intl` with values from intl config props', () => {
+    areIntlLocalesSupported.mockClear();
+    areIntlLocalesSupported.mockReturnValue(true);
```
Comment (emmenko) [inline test/unit/components/provider.tsx]: We just return `true`, effectively whitelisting all the languages. I think for this kind of test this is totally fine.

**Comment by longlho** [comment]
overall LGTM! Ready to merge once build if fixed. Thanks a lot for your contributions!

**Comment by emmenko** [comment]
Awesome, glad to help!

**Comment by emmenko** [comment]
Hmm wait, if I run `npm test` locally I get this:

```
❯ npm test                                     

> react-intl@3.1.6 test /Users/emmenko/dev/src/react-intl
> cross-env NODE_ICU_DATA=./node_modules/full-icu jest --coverage --verbose

node: could not initialize ICU (check NODE_ICU_DATA or --icu-data-dir parameters)
npm ERR! Test failed.  See above for more details.
```

I've been using `npm test:watch` to work on the PR...did I miss some step? 🤔

**Comment by emmenko** [comment]
I guess that might explain the problem I had with https://github.com/formatjs/react-intl/pull/1413#discussion_r312929628

**Comment by emmenko** [comment]
Ah ok, I had to `npm rebuild` as I switched node versions after installing the deps the first time 😅

**Comment by emmenko** [comment]
Alright, I force pushed and removed the `jest.mock('intl-locales-supported');` workaround.

File: package.json [hunk 9]
```diff
@@ -112,7 +112,7 @@
     "release": "standard-version",
     "test:all": "npm run lint && npm run format && npm run test",
     "test:perf": "cross-env NODE_ENV=production babel-node test/perf",
-    "test:watch": "jest --watch",
+    "test:watch": "cross-env NODE_ICU_DATA=./node_modules/full-icu jest --watch",
```
Comment (emmenko) [inline package.json]: This is what tricked me. Now the polyfilled data is loaded also when running jest in watch mode.

Comment (longlho) [reply package.json]: then u prob don't need to mock `intl-locales-supported` right?

**Comment by longlho** [comment]
Ah Node 12 fails bc it comes w/ native `Intl.RelativeTimeFormat` so u might just to adjust the err message expectation

**Comment by DragonRaider5** [comment]
The reason why injectIntl was still being used is because it allows for more flexibility in the SCU.
Just wanted to throw this out there before the merge.

**Comment by longlho** [comment]
@DragonRaider5 	can u provide more details? I'd love to understand that use case more.

**Comment by emmenko** [comment]
> The reason why injectIntl was still being used is because it allows for more flexibility in the SCU.

The only place where we currently use sCU is the `message` component. I don't see a problem with not using `injectIntl` to be honest, because if `intl` changes, React will update the context, causing a new re-render anyway.

Unless I'm missing some other point...
